Octave pkg install mkoctfile and build-from-source failures

**Please note we will close your issue without comment if you delete, do not read or do not fill out the issue checklist below and provide ALL the requested information. If you repeatedly fail to use the issue template, we will block you from ever submitting issues to Homebrew again.**

- [ x] are reporting a bug others will be able to reproduce and not asking a question. If you're not sure or want to ask a question do so on our Discourse: https://discourse.brew.sh
- [ ] have a problem with `brew install` (or `upgrade`, `reinstall`) a single, official formula (not cask)? If it's a general `brew` problem please file this issue at Homebrew/brew: https://github.com/Homebrew/brew/issues/new. If it's a `brew cask` problem please file this issue at https://github.com/caskroom/homebrew-cask/issues/new. If it's a tap (e.g. Homebrew/homebrew-php) problem please file this issue at the tap.
- [ x] ran `brew update` and can still reproduce the problem?
- [x ] ran `brew doctor`, fixed all issues and can still reproduce the problem?
- [ x] ran `brew gist-logs <formula>` (where `<formula>` is the name of the formula that failed) and included the output link?
- [ ] if `brew gist-logs` didn't work: ran `brew config` and `brew doctor` and included their output with your issue?

To help us debug your issue please explain:
- What you were trying to do (and why)
- What happened (include command output)
- What you expected to happen
- Step-by-step reproduction instructions (by running `brew install` commands)

# Formula additions or changes
To get formulae added or changed in Homebrew please file a [Pull Request](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)
We will close issues requesting formulae changes.

First, I encountered the same issue as was reported in 

[Homebrew/homebrew-science] ld: library not found -lgfortran, during pkg install -forge control in octave 4.2.0_2 (#4972)

While installing octave pkg control using 

pkg install -forge control
 
...

ld: warning: directory not found for option '-L/usr/local/Cellar/gcc/7.3.0/lib/gcc/7/gcc/x86_64-apple-darwin17.3.0/7.3.0'
ld: warning: directory not found for option '-L/usr/local/Cellar/gcc/7.3.0/lib/gcc/7/gcc/x86_64-apple-darwin17.3.0/7.3.0/../../..'
ld: library not found for -lgfortran
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [__control_slicot_functions__.oct] Error 1


brew info octave
octave: stable 4.2.1 (bottled), HEAD
High-level interpreted language for numerical computing
https://www.gnu.org/software/octave/index.html
/usr/local/Cellar/octave/4.2.1_10 (2,182 files, 37.6MB)
  Poured from bottle on 2018-01-20 at 08:44:01
/usr/local/Cellar/octave/4.2.1_11 (2,182 files, 37.6MB) *
  Poured from bottle on 2018-02-23 at 16:16:10
From: https://github.com/Homebrew/homebrew-core/blob/master/Formula/octave.rb
==> Dependencies
Build: automake ✔, autoconf ✔, gnu-sed ✔, pkg-config ✔
Required: arpack ✔, epstool ✔, fftw ✔, fig2dev ✔, fltk ✔, fontconfig ✔, freetype ✔, gcc ✔, ghostscript ✔, gl2ps ✔, glpk ✔, gnuplot ✔, graphicsmagick ✔, hdf5 ✔, libsndfile ✔, libtool ✔, pcre ✔, portaudio ✔, pstoedit ✔, qhull ✔, qrupdate ✔, readline ✔, suite-sparse ✔, veclibfort ✔
==> Requirements
Optional: java >= 1.6 ✘
==> Options
--with-java
	Build with java support
--HEAD
	Install HEAD version

So, to fix this, I tried reinstalling octave from source as was done in #4972.

brew upgrade
brew uninstall octave
Uninstalling /usr/local/Cellar/octave/4.2.1_11... (2,182 files, 37.6MB)
octave 4.2.1_10 1 is still installed.
brew install --build-from-source octave
Updating Homebrew...
==> Auto-updated Homebrew!
Updated 1 tap (caskroom/cask).
No changes to formulae.

==> Downloading https://ftp.gnu.org/gnu/octave/octave-4.2.1.tar.gz
######################################################################## 100.0%
==> ./configure --prefix=/usr/local/Cellar/octave/4.2.1_11 --disable-silent-rules --enable-link-all-dependencies --enable-shared --disable-static --d
==> make all
Last 15 lines from /Users/sunil/Library/Logs/Homebrew/octave/02.make:
all

preserving existing HG-ID file
CDPATH="${ZSH_VERSION+.}:" && cd . && /bin/sh /private/tmp/octave-20180312-37007-dten9x/octave-4.2.1/build-aux/missing aclocal-1.15 -I m4
/private/tmp/octave-20180312-37007-dten9x/octave-4.2.1/build-aux/missing: line 81: aclocal-1.15: command not found
WARNING: 'aclocal-1.15' is missing on your system.
         You should only need it if you modified 'acinclude.m4' or
         'configure.ac' or m4 files included by 'configure.ac'.
         The 'aclocal' program is part of the GNU Automake package:
         <http://www.gnu.org/software/automake>
         It also requires GNU Autoconf, GNU m4 and Perl in order to run:
         <http://www.gnu.org/software/autoconf>
         <http://www.gnu.org/software/m4/>
         <http://www.perl.org/>
make: *** [aclocal.m4] Error 127

READ THIS: https://docs.brew.sh/Troubleshooting

brew gist-logs octave
https://gist.github.com/b31a8de38ca7f706269216c45f05cefe

--- a/Formula/octave.rb
+++ b/Formula/octave.rb
@@ -99,7 +99,12 @@ class Octave < Formula
 
     args << "--disable-java" if build.without? "java"
 
-    system "./bootstrap" if build.head?
+    if build.head?
+      system "./bootstrap"
+    else
+      system "autoreconf", "-fiv"
+    end
+
     system "./configure", *args
     system "make", "all"
     system "make", "install"

